src/lock.rs

```diff
@@ -0,0 +1,174 @@
+// (c) Copyright 2023 Helsing GmbH. All rights reserved.
+
+use std::{collections::HashMap, sync::Arc};
+
+use eyre::{ensure, Context};
+use ring::digest;
+use semver::{Version, VersionReq};
+use serde::{Deserialize, Serialize};
+use tokio::fs;
+
+use crate::{
+    manifest::{Dependency, DependencyManifest},
+    package::{Package, PackageName},
+};
+
+pub const LOCKFILE: &str = "Proto.lock";
+
+/// Captures immutable metadata about a given package
+///
+/// It is used to ensure that future installations will use the exact same dependencies.
+#[derive(Serialize, Deserialize, Clone)]
+pub struct LockedPackage {
+    pub name: PackageName,
+    pub checksum: String,
+    pub repository: String,
+    pub version: Version,
+    pub dependencies: Vec<PackageName>,
```

Comment:
It's possible I've misunderstood how buffrs has evolved. As I've understood it, Buffrs doesn't do dependency resolution - rather it recompiles and vendors packages - such that you get some package duplication.

This would imply that dependencies would need to depend on both PackageName and Version. Is this still how buffrs works?

Reply:
Ah, currently we only support exact versions, so if a dependency is required by two packages in the graph it will error out unless both request the same version. We'll need to make some changes to the `DependencyGraph` algorithm later to handle actual version resolution.

Reply:
No - buffrs does dependency resolution! We introduced this in v0.4

Reply:
I'd actually value your input on whether this is headed in the right direction for that. My naive resolution algorithm would look somewhat like this:
1. When a dependency is first required (eg, it is included in the manifest), we assume the latest allowed version and recurse its sub-dependencies at that version.
2. If a sub-dependency (or another dependency) later is found to require the same package again, we add a constraint, recalculate the latest allowed version and schedule the sub-dependencies for traversal again (if they changed).

There might be edge cases leading to cycles which we'd need to detect, but I think this should work in general. I could be wrong... Haven't thought too deeply about it yet, but if I'm right the current data structure should work as is.

Reply:
Then is it worth leaving room for the dependencies in the lockfile to be dependency@versions? Rather than a single version thing? Or is it intended that buffrs unifies into a single tree permanently?

Reply:
I didn't see andre's reply.

My understanding of how you do simple dependency resolution in a statically typed language (which this is!) is:

1. You always pick the max version of what you've seen.
2. You always use that version, user gets no choice over the matter.

And if there's breakage, there's breakage and dependencies can't be used together. The alternatives are:

1. Do what python does - allow the user to specify dependency ranges, carets, tildes, all kinds of crap. The downside of this is that it's easy for one package to specify incompatibility anyway and so you end up with deps that can't be used together all the time even though the code works fine. This is good for Python as Python is fundamentally an untyped language and so you need the metadata.
2. Do what Rust does - allow multiple crates to be in use at any one time.
3. Do what Java does - if you make lots of breaking changes, that's your bad, don't do that.

What Python does is basically unsuitable for buffrs and unnecessary. The difference is whether you think you'll go down the java path or the rust path. If you might go down the rust path, probably worth containing the version in the lockfile spec.

An additional comment here: there might be no need to actually store the dependencies in the lockfile entry. In Python or Rust it _can_ be useful because you can put your feature flags and whatnot into the lockfile so you can filter down what you need to download based on platform and configuration. Whereas buffrs doesn't really have this. Without feature flags, platform specific considerations etc, you might just want to store the list of dependencies.

Reply:
Last comment on lockfile structure, but i promise, this one is important! At Palantir, we did java lockfiles https://github.com/palantir/gradle-consistent-versions/blob/develop/versions.lock - and a thing we noticed was that with a naive lockfile, it's easy to end up with semantic merge conflicts which cause the lockfile to be broken without generating a git conflict.

Effectively, you want unrelated concurrent updates to not generate git conflicts, but for related concurrent updates to generate git conflicts.

A simple (slightly contrived, but I promise this comes up in the real world quite a bit - I just can't remember the exact case) example of this would be:

```
x@1.1 depends on a@1.1 which depends on c@1.4
x@1.1 depends on b@1.2 which depends on c@1.4
a@1.2 removes the dependency on c
b@1.3 removes the dependency on c
```

and now if I had concurrent PRs to update `a` and `b`, neither of my PRs would remove the dependency on `c` from the lockfile - but after they merge, `c` should no longer be in the lockfile.

At palantir the goal was to have a very compact lockfile, with the hope being that instead of reviewing the list of dependency updates, you'd review the lockfile to see what had changed. So there the solution was that you'd build a hash of your dependents and stick that on the line.

So in the above case, the c dependency would have hash(a, b) next to it, and a change to either a or b would change the line referring to c.

Reply:
the tl;dr of the above being

- maybe discard the dependency graph from the lockfile?
- maybe instead put a hash of it on the consumers? not useful for dependency resolution but useful for causing merge conflicts

Reply:
> maybe discard the dependency graph from the lockfile?

Rationale for keeping it in is that it could be useful for local diagnostics. Without that information we'd need to use the registry to resolve the dependency tree for inspection (for some `explain` command or similar).

> maybe instead put a hash of it on the consumers? not useful for dependency resolution but useful for causing merge conflicts

I think we may get by with just a count of dependants. I'm introducing a lockfile version field so that we can handle future migrations if necessary.

Reply:
Ok, implemented the dependants count.